spec/javascripts/packages/document-capture/components/document-capture-spec.jsx

```diff
@@ -275,4 +294,106 @@ describe('document-capture/components/document-capture', () => {
   });
+
+  it('renders async upload pending progress', async () => {
+    const statusChecks = 3;
+    let remainingStatusChecks = statusChecks;
+    sandbox.stub(window, 'fetch');
+    const upload = sinon.stub().callsFake((payload, { endpoint }) => {
+      switch (endpoint) {
+        case 'about:blank#upload':
+          expect(payload).to.have.keys([
+            'front_image_iv',
+            'front_image_url',
+            'back_image_iv',
+            'back_image_url',
+            'selfie_image_iv',
+            'selfie_image_url',
+          ]);
+
+          return Promise.resolve({ success: true, isPending: true });
+        case 'about:blank#status':
+          expect(payload).to.be.empty();
+
+          return Promise.resolve({ success: true, isPending: Boolean(remainingStatusChecks--) });
+        default:
+          throw new Error();
+      }
+    });
+    const key = await window.crypto.subtle.generateKey(
+      {
+        name: 'AES-GCM',
+        length: 256,
+      },
+      true,
+      ['encrypt', 'decrypt'],
+    );
+
+    const { getByLabelText, getByText, getAllByText, findAllByText } = baseRender(
+      <UploadContextProvider
+        endpoint="about:blank#upload"
+        statusEndpoint="about:blank#status"
+        backgroundUploadURLs={{
+          front: 'about:blank#front',
+          back: 'about:blank#back',
+          selfie: 'about:blank#back',
+        }}
+        backgroundUploadEncryptKey={key}
+        upload={upload}
+      >
+        <AcuantProvider sdkSrc="about:blank">
+          <DocumentCapture isAsyncForm />
+        </AcuantProvider>
+      </UploadContextProvider>,
+    );
+
+    initialize({ isCameraSupported: false });
+    window.AcuantPassiveLiveness.startSelfieCapture.callsArgWithAsync(0, '');
+
+    const continueButton = getByText('forms.buttons.continue');
+    userEvent.click(continueButton);
+    await findAllByText('simple_form.required.text');
+    userEvent.upload(
+      getByLabelText('doc_auth.headings.document_capture_front'),
+      new window.File([''], 'upload.png', { type: 'image/png' }),
+    );
+    userEvent.upload(
+      getByLabelText('doc_auth.headings.document_capture_back'),
+      new window.File([''], 'upload.png', { type: 'image/png' }),
+    );
+    await waitFor(() => expect(() => getAllByText('simple_form.required.text')).to.throw());
+    userEvent.click(continueButton);
+
+    const submitButton = getByText('forms.buttons.submit.default');
+    userEvent.click(submitButton);
+    await findAllByText('simple_form.required.text');
+    const selfieInput = getByLabelText('doc_auth.headings.document_capture_selfie');
+    fireEvent.click(selfieInput);
+    await waitFor(() => expect(() => getAllByText('simple_form.required.text')).to.throw());
+    userEvent.click(submitButton);
+
+    return new Promise((resolve) => {
+      const form = document.createElement('form');
+      form.className = 'js-document-capture-form';
+      document.body.appendChild(form);
```

Comment:
is this a case where we would just want to use react components to render a form? Or is it more advantageous to do this manually?

```
<form class="js-document-capture-form" onsubmit={submitFn}>
</form>
```

Reply:
> is this a case where we would just want to use react components to render a form?

Do you mean within the document capture application, or as part of this test code?

But yeah, since we've repeated this a few times already, I think there could be a better way to reuse this common logic. Possibly something integrated as part of the `useCleanDOM` (`initialize` callback) or part of or complementing the existing `useAcuant` test helper.

Reply:
I meant the test code

Reply:
> I meant the test code

👍 

I've an idea for a bit of a broader reorganization of the document capture test setup. I'll plan to open a pull request immediately following this.

Reply:
>I'll plan to open a pull request immediately following this.

See: #4358